src/server/transaction/server/driver.go

```diff
@@ -197,26 +198,34 @@ func (d *driver) runTransaction(txnCtx *txncontext.TransactionContext, info *tra
 
 func (d *driver) finishTransaction(ctx context.Context, txn *transaction.Transaction) (*transaction.TransactionInfo, error) {
 	info := &transaction.TransactionInfo{}
-	if err := d.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
-		err := d.transactions.ReadOnly(ctx).Get(txn.ID, info)
-		if err != nil {
-			return err
-		}
-		info, err = d.runTransaction(txnCtx, info)
-		if err != nil {
-			return err
+	return d.updateTransaction(ctx, func() (*transaction.TransactionInfo, error) {
+		if err := d.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+			err := d.transactions.ReadWrite(txnCtx.SqlTx).Get(txn.ID, info)
+			if err != nil {
+				return err
+			}
+			info, err = d.runTransaction(txnCtx, info)
+			if err != nil {
+				if col.IsErrTransactionConflict(err) {
+					// a pipeline changed, and we need to update the transaction data
```

Comment:
This _could_ happen for other reasons in the future, maybe make this comment more general or just list that as an example.